File: graph/dfs_recursive/cpp/DFS.cpp
#include "DFS.hpp"

#include <functional>
#include <vector>

using namespace std;

vector<int> DFS(const vector<vector<int>>& graph, int start) {
    vector<int> order;
    vector<bool> visited(graph.size(), false);

    function<void(int)> dfs = [&](int node) {
        visited[node] = true;
        order.push_back(node);

        for (const int neighbor : graph[node]) {
            if (!visited[neighbor]) {
                dfs(neighbor);
            }
        }
    };

    dfs(start);
    return order;
}

File: graph/dfs_recursive/rust/src/DFS.rs
#[allow(non_snake_case)]
pub fn DFS(graph: &[Vec<usize>], start: usize) -> Vec<usize> {
    let mut order = Vec::new();
    let mut visited = vec![false; graph.len()];

    fn dfs(
        node: usize,
        graph: &[Vec<usize>],
        visited: &mut [bool],
        order: &mut Vec<usize>,
    ) {
        visited[node] = true;
        order.push(node);

        for &neighbor in &graph[node] {
            if !visited[neighbor] {
                dfs(neighbor, graph, visited, order);
            }
        }
    }

    dfs(start, graph, &mut visited, &mut order);
    order
}
